numpy/ma/core.py

```diff
@@ -4572,21 +4590,25 @@ def sum(self, axis=None, dtype=None, out=None):
         newmask = _check_mask_axis(_mask, axis)
         # No explicit output
         if out is None:
-            result = self.filled(0).sum(axis, dtype=dtype)
+            result = self.filled(0).sum(axis, dtype=dtype, keepdims=keepdims)
             rndim = getattr(result, 'ndim', 0)
             if rndim:
                 result = result.view(type(self))
-                result.__setmask__(newmask)
+                if (newmask is nomask):
+                    result.__setmask__(nomask)
+                else:
+                    result.__setmask__(newmask.reshape(result.shape))
             elif newmask:
                 result = masked
             return result
         # Explicit output
-        result = self.filled(0).sum(axis, dtype=dtype, out=out)
+        result = self.filled(0).sum(axis, dtype=dtype, out=out,
+                                    keepdims=keepdims)
         if isinstance(out, MaskedArray):
             outmask = getattr(out, '_mask', nomask)
             if (outmask is nomask):
                 outmask = out._mask = make_mask_none(out.shape)
-            outmask.flat = newmask
+            outmask[()] = newmask.reshape(result.shape)
```

Comment:
hmm you're right using `.flat` works. I was worried that a non-1d newmask would cause problems.

I still think `outmask[...] = ` is better (thanks for explaining `()` vs `...`). I guess I am biased against `.flat` after reading some other devs say they didn't like it!